File: 백준/Gold/17281. ⚾/⚾.cc
#include <iostream>
#include <algorithm>
#include <cstring>
#include <queue>
#include <cmath>

using namespace std;
#define MAX 1001
#define endl '\n'

int N;
int answer;
int Game[51][51];
bool Select[10];
int Order[10];

void Play() {
	int score = 0;
	int start_player = 1;
	int base_state[4];
	for (int i = 1;i <= N;i++) {
		int out_cnt = 0;
		bool next_inning = false;
		memset(base_state, 0, sizeof(base_state));
		while(1) {
			for (int j = start_player;j < 10;j++) {
				int player = Game[i][Order[j]];
				if (player == 0)
					out_cnt++;
				else if(player==1) {
					for (int F = 3;F >= 1;F--) {
						if (base_state[F] == 1) {
							if (F == 3) {
								base_state[F] = 0;
								score++;
							}
							else {
								base_state[F + 1] = 1;
								base_state[F] = 0;
							}
						}
					}
					base_state[1] = 1;
				}
				else if (player == 2) {
					for (int F = 3;F >= 1;F--) {
						if (base_state[F] == 1) {
							if (F == 3 || F==2) {
								base_state[F] = 0;
								score++;
							}
							else {
								base_state[F + 2] = 1;
								base_state[F] = 0;
							}
						}
					}
					base_state[2] = 1;
				}
				else if (player == 3) {
					for (int F = 3;F >= 1;F--) {
						if (base_state[F] == 1) {
							base_state[F] = 0;
							score++;
						}
					}
					base_state[3] = 1;
				}
				else {
					for (int F = 3;F >= 1;F--) {
						if (base_state[F] == 1) {
							base_state[F] = 0;
							score++;
						}
					}
					score++;
				}
				if (out_cnt == 3) {
					start_player = j + 1;
					if (start_player == 10)
						start_player = 1;
					next_inning = true;
					break;
				}
			}
			if (next_inning == true)
				break;
			start_player = 1;
		}
	}
	answer = max(answer, score);
}

void dfs(int cnt) {
	if (cnt == 10) {
		Play();
		return;
	}
	for (int i = 1;i <= 9;i++) {
		if (Select[i])
			continue;
		Select[i] = true;
		Order[i] = cnt;
		dfs(cnt + 1);
		Select[i] = false;
	}
}

int main(void) {
	cin >> N;
	for (int i = 1;i <= N;i++) {
		for (int j = 1;j <= 9;j++) {
			cin >> Game[i][j];
		}
	}
	Select[4] = true;
	Order[4] = 1;
	dfs(2);
	cout << answer;
}

File: 백준/Gold/17281. ⚾/⚾.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.StringTokenizer;

class Main {
    static int N, answer = 0;
    static int[][] game;
    static boolean[] isSelect = new boolean[10];
    static int[] order = new int[10];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        game = new int[N + 1][10];
        for (int i = 1; i <= N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 1; j <= 9; j++) {
                game[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        isSelect[4] = true;
        order[4] = 1;
        select(2);
        System.out.println(answer);
    }

    static void select(int cnt) {
        if (cnt == 10) {
//            System.out.println(Arrays.toString(order));
            play();
            return;
        }
        for (int i = 1; i <= 9; i++) {
            if (!isSelect[i]) {
                isSelect[i] = true;
                order[i] = cnt;
                select(cnt + 1);
                isSelect[i] = false;
            }
        }
    }

    static void play() {
        int score = 0;
        int playerNum = 1;
        int[] state;
        for (int i = 1; i <= N; i++) {
            int outCnt = 0;
            boolean isNext = false;
            state = new int[4];
            while (true) {
                for (int j = playerNum; j <= 9; j++) {
                    int result = game[i][order[j]];
                    if (result == 0) {
                        outCnt++;
                    } else if (result == 1) {
                        for (int idx = 3; idx >= 1; idx--) {
                            if (state[idx] == 1) {
                                if (idx == 3) {
                                    state[idx] = 0;
                                    score++;
                                } else {
                                    state[idx + 1]++;
                                    state[idx] = 0;
                                }
                            }
                        }
                        state[1] = 1;
                    } else if (result == 2) {
                        for (int idx = 3; idx >= 1; idx--) {
                            if (state[idx] == 1) {
                                if (idx == 3 || idx == 2) {
                                    state[idx] = 0;
                                    score++;
                                } else {
                                    state[idx + 2]++;
                                    state[idx] = 0;
                                }
                            }
                        }
                        state[2] = 1;
                    } else if (result == 3) {
                        for (int idx = 3; idx >= 1; idx--) {
                            if (state[idx] == 1) {
                                state[idx] = 0;
                                score++;
                            }
                        }
                        state[3] = 1;
                    } else {
                        for (int idx = 3; idx >= 1; idx--) {
                            if (state[idx] == 1) {
                                state[idx] = 0;
                                score++;
                            }
                        }
                        score++;
                    }
                    if (outCnt == 3) {
                        playerNum = j + 1;
                        if (playerNum == 10)
                            playerNum = 1;
                        isNext = true;
                        break;
                    }
                }
                if (isNext)
                    break;
                playerNum = 1;
            }
        }
        answer = Math.max(answer, score);
    }
}